source/Detector/Amptek/AmptekSDD123Server.cpp

```diff
@@ -0,0 +1,522 @@
+#include "AmptekSDD123Server.h"
+
+#include <QtNetwork>
+#include <QCoreApplication>
+#include <QDebug>
+
+#include "AmptekEventDefinitions.h"
+#include "util/AMErrorMonitor.h"
+
+/* Notes on how to convert back and forth
+QString dataString = "4d4341433d3f3b";
+QString plainString = "MCAC=?;";
+qDebug() << "Normal " << dataString << plainString;
+qDebug() << "In hex " << dataString << plainString.toAscii().toHex();
+qDebug() << "In ASCII " << QString(QByteArray::fromHex(dataString.toAscii())) << plainString;
+*/
+
+AmptekSDD123Server::AmptekSDD123Server(AmptekSDD123ConfigurationMap *configurationMap, const QString &serverName, QObject *parent) :
+	QObject(parent) , currentRequestPacket_(-1, QString("NONE")), currentSyncPacket_(-1, QString("NONE")), timedOutPacket_(-1, QString("NONE"))
+{
+	configurationMap_ = configurationMap;
+	serverName_ = serverName;
+	spectrumPacketReceiver_ = 0;
+	confirmationPacketReceiver_ = 0;
+	requestIntervalTimer_ = -1;
+
+	if(configurationMap_){
+		stillReceivingResponseDatagram_ = false;
+		socketLocallyBusy_ = false;
+		detectorUnavailable_ = false;
+
+		currentRequestTime_.start();
+		requestPacketTimer_ = new QTimer(this);
+		requestPacketTimer_->setSingleShot(true);
+		syncPacketTimer_ = new QTimer(this);
+		syncPacketTimer_->setSingleShot(true);
+
+		packetIDCounter_ = 27;
+		hasReplyReady_ = false;
+
+		timeOutIDCounter_ = -27;
+
+		requestSpectrumTime_ = new QTime();
+		requestSpectrumTime_->start();
+		lastRequestSpectrumTime_ = new QTime();
+		lastRequestSpectrumTime_->start();
+		replySpectrumTime_ = new QTime();
+		replySpectrumTime_->start();
+
+		udpDetectorSocket_ = new QUdpSocket(this);
+		bool boundUDP = udpDetectorSocket_->bind(configurationMap_->localAddress(), 10001, QUdpSocket::ShareAddress);
+		if (!boundUDP)
+			AMErrorMon::alert(this, AMPTEK_SERVER_ALERT_FAILED_TO_BOUND_UDP, QString("Amptek Server failed to bound to the detector via %1:%2").arg(configurationMap_->localAddress().toString()).arg(10001));
+
+		connect(udpDetectorSocket_, SIGNAL(readyRead()), this, SLOT(processPendingDatagrams()));
+		connect(this, SIGNAL(responsePacketReady()), this, SLOT(onResponsePacketReady()));
+	}
+}
+
+QList<double> AmptekSDD123Server::droppedPacketStatistics() const
```

Comment:
Do we only ever care about the last 30 seconds?  Would it make sense to have a version that took in a value and returned the statistics in that interval?